pkg/kubelet/kubelet_pods.go

```diff
@@ -1564,6 +1571,51 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			continue
 		}
 		status := convertContainerStatus(cStatus)
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			if status.State.Running != nil {
+				var requests, limits v1.ResourceList
+				// Set initial limits from container's spec upon transition to Running state
+				// For cpu & memory, limits queried from runtime via CRI always supercedes spec.limit
+				// For ephemeral-storage, a running container's status.limit equals spec.limit
+				if containerResources[cName].Limits != nil {
+					// oldStatus should always exist if container state is running? TODO: Verify this.
+					oldStatus, oldStatusFound := oldStatuses[cName]
+					limits = make(v1.ResourceList)
+					determineResourceLimits := func(rName v1.ResourceName) {
+						if res, found := cStatus.ResourceLimits[rName]; found {
+							limits[rName] = res.DeepCopy()
+							return
+						}
+						if oldStatusFound {
+							if oldStatus.State.Running == nil || status.ContainerID != oldStatus.ContainerID {
+								if res, exists := containerResources[cName].Limits[rName]; exists {
+									limits[rName] = res.DeepCopy()
+								}
+							} else {
+								if oldStatus.Resources.Limits != nil {
+									if res, exists := oldStatus.Resources.Limits[rName]; exists {
+										limits[rName] = res.DeepCopy()
+									}
+								}
+							}
+						}
+					}
+					determineResourceLimits(v1.ResourceCPU)
+					determineResourceLimits(v1.ResourceMemory)
+					if ephemeralStorage, found := containerResources[cName].Limits[v1.ResourceEphemeralStorage]; found {
+						limits[v1.ResourceEphemeralStorage] = ephemeralStorage.DeepCopy()
+					}
+				}
+				// A running container's status.requests always comes from spec.resourcesAllocated
```

Comment:
Shouldn't requests also come from the container runtime?  Requests are used for CPU shares IIRC.

Reply:
They should but I hit a corner-case I was unable to resolve. For millicpu requests of 1 vs 2, they both get a floor cpu.shares of 2. So when runtime reports back 2, I cannot tell whether it is 1m or 2m. Is it OK to report 2m even though user requested 1m? Maybe it is.. the container is going to get minimum shares of 2 regardless of 1m or 2m ask.

Even if I didn't do the above, there is another bug here - I should be updating ContainerStatus.Resources.Requests only after successful updateContainerLimits for all containers. I'll fix that, and revisit if I can use cpu.shares reported by runtime to reverse-compute cpu requests.

Reply:
Switched to getting Requests from cpu.shares in commit [2d88cb1](https://github.com/vinaykul/kubernetes/pull/1/commits/2d88cb142cb73073aa187edb2c8b6535ca416b8c). Seems to work fine.